### PR TITLE
introduce interactive mode and new process model

### LINKED ISSUE #36: Test status indicator issue
When there's a failing test, the indicator turns red:

test js - first-agenda 2016-12-05 11-41-18

When I fix it, the comment disappears, but the indicator is still red:

test js - first-agenda 2016-12-05 11-41-38

### LINKED ISSUE #119: [Feature request] Ability to just run the last changed test file
Just had a case where I was working on file `foo.js` and `foo.spec.js`. Foo is used in a bunch of places, so upon changing it Jest correctly figures out that it has 10 spec files that it should run. This is great most of the time, but it means that I have to wait longer to know if `foo.spec.js` has passed or not. An option to turn on a "single file" mode where it only runs the last changed spec file on each change would be quite useful.

I'm quite happy to attempt some work on this if you think it's a reasonable addition :)

### LINKED ISSUE #176: [Feature request] Add ability to force rerun of all tests
Would be nice to have a feature (a command for instance) to rerun all tests without stopping and restarting the extension.

### LINKED ISSUE #308: Ext won't refresh after code change
I purposefully broke one of my tested methods, and this ext gives me an error. I undo my changes, and the ext still gives me an error. 

Is there any way to force a refresh of the test results so that the error disappears, after it's been fixed?

### LINKED ISSUE #408: Tests are failing when running with plugin but working fine with command line
### Environment

1. `node -v`: v10.13.0
1. `npm -v`: 6.4.1
1. `npm ls jest` or `npm ls react-scripts` (if you haven’t ejected):
$ npm ls jest
jest_vscode_failing_example_app@1.0.0 C:\MyExperiments\jest_vscode_failing_example_app
`-- jest@22.4.4
(also tried 23.6.0 with the same results)
1. your vscode-jest settings if customized:
    - jest.pathToJest? null
    - jest.pathToConfig? ""
    - anything else that you think might be relevant? no

1. Operating system: Windows 10 Version 1803 (OS Build 17134.407)

### Prerequisite
- are you able to run jest test from command line? yes
- how do you run your tests from command line? (for example: `npm run test` or `node_modules/.bin/jest`) npm run test

### Steps to Reproduce

https://github.com/unfalse/vscode_jest_fail

### Relevant Debug Info

This is the messages from the output window.
```
--watch is not supported without git/hg, please use --watchAll 

  console.warn node_modules/react-addons-test-utils/index.js:31
    Warning: ReactTestUtils has been moved to react-dom/test-utils. Update references to remove this warning.


FAIL src/index.test.js

  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    SyntaxError: C:\MyExperiments\jest_vscode_failing_example_app\src\index.test.js: Unexpected token (7:30)

       5 | describe('App: ', () => {
       6 |     it('renders withour crash', () => {
    >  7 |         var wrapper = shallow(<App/>);
         |                               ^
       8 |         expect(wrapper.find('div').length).toEqual(1);
       9 |     });
      10 | });

      at Parser.raise (node_modules/@babel/parser/lib/index.js:4028:15)
      at Parser.unexpected (node_modules/@babel/parser/lib/index.js:5343:16)
      at Parser.parseExprAtom (node_modules/@babel/parser/lib/index.js:6432:20)
      at Parser.parseExprSubscripts (node_modules/@babel/parser/lib/index.js:6019:21)
      at Parser.parseMaybeUnary (node_modules/@babel/parser/lib/index.js:5998:21)
      at Parser.parseExprOps (node_modules/@babel/parser/lib/index.js:5907:21)
      at Parser.parseMaybeConditional (node_modules/@babel/parser/lib/index.js:5879:21)
      at Parser.parseMaybeAssign (node_modules/@babel/parser/lib/index.js:5826:21)
      at Parser.parseExprListItem (node_modules/@babel/parser/lib/index.js:7111:18)
      at Parser.parseCallExpressionArguments (node_modules/@babel/parser/lib/index.js:6227:22)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        4.518s
Ran all test suites.
```

### Expected Behavior

vscode-jest is not failing on tests.

### Actual Behavior

vscode-jest failing on tests but `npm run test` works fine.

---

### LINKED ISSUE #530: How to run test cases of only the file currently being edited.
### Environment

1. `node -v`: [fill]
1. `npm -v`: [fill]
1. `npm ls jest` or `npm ls react-scripts` (if you haven’t ejected): [fill]
1. your vscode-jest settings if customized:
    - jest.pathToJest? [fill]
    - jest.pathToConfig? [fill]
    - anything else that you think might be relevant? [fill]

1. Operating system: [fill]

### Prerequisite
- are you able to run jest test from command line? [fill]
- how do you run your tests from command line? (for example: `npm run test` or `node_modules/.bin/jest`) [fill]

### Steps to Reproduce

[fill]

### Relevant Debug Info

[fill]

### Expected Behavior

[fill]

### Actual Behavior

[fill]

--- 

The fastest (and the most fun) way to resolve the issue is to submit a pull-request yourself. If you are interested, feel free to check out the [contribution guide](https://github.com/jest-community/vscode-jest/CONTRIBUTING.md), we look forward to seeing your PR...

### LINKED ISSUE #574: FR: workflow improvements when snapshot matching fails
I am really enjoying the extension. I have a couple of requests that I think will help the workflow when snapshots fail. 
I notice that there is a Jest: ... section in the status bar and it shows when the tests are running. This is fantastic. 

Here is my workflow when snapshots fail. It is a bit clunky. This is what happens now:

- I make a change in my vue file and that triggers a unit test run
- 20 seconds later a notification pops up saying that the snapshot is out of day (I use TS, hense the 20 second to run a test)
- I click on the "Refresh snapshot" button
- often a second notification will come up quickly with another refresh button. I am not sure if I am supposed to hit that or not
- I sit there staring at the bottom of the screen for the notification that says snapshots have been updated. 
- I re-save my unit test file to trigger the test run again.

When each test run and snapshot generation takes 20 seconds (not your fault), this workflow becomes very time consuming. Here are some things I would love to see:

- the status bar area: it would be nice if it showed when snapshot generation was occuring
   - there are times when I don't know if it is generating snapshots in the background
   - I also don't know if saving the unit test file again to kick off a new run will interfere with the snapshot generation or not
- I assume that when I get multiple dialogs with the refresh button, that multiple test files have been run and have snapshot mismatches. It would be nice if the notification box told me which file was run and had failures.
- It would be nice if there was some way to kick off snapshot update without having to wait for a failed test run first. Maybe beside the "view snapshot" button, there could be a "regenerate snapshot" button. 

Thanks for listening

### LINKED ISSUE #613: Support disabling watch mode
There doesn't seem to be a way to disable the watch mode. With watch mode on the tests are always running whenever there is a save action on any file in the workspace. When we have a decent number of tests in the workspace it takes some time to finish all tests, which causes a high cpu load on my machine (observed by just listening to the fan noise).

Is it possible to disable the watch mode so that we have the ability to control when the tests are being run?

### LINKED ISSUE #668: Watch doesn't update test result when problem is fixed (only on Error)
### Environment

1. `node -v`: v14.5.0
1. `npm -v`: 6.14.5
1. `npm ls jest` or `npm ls react-scripts` (if you haven’t ejected): jest@26.6.3
1. Operating system: Windows_NT x64 10.0.19042
2. VS Code: 1.53.2

### Prerequisite
- are you able to run jest test from command line? YES
- how do you run your tests from command line? `npm run jest`

### Steps to Reproduce

1. I have working (passing) test unit, I introduce a typo, save file => error shown in PROBLEMS tab + error markers in the file
2. I fix typo, save file => error still in the PROBLEMS tab +  error markers in the file
3. I introduce new typo => new (correct) error in the PROBLEMS tab +  error markers in the file
4. I fix new typo, save file => error still in the PROBLEMS tab + error markers in the file
5. restarting jest watcher clears PROBLEMS tab + success markers in the file until next error.

### Relevant Debug Info

It appeared on async test.

### Expected Behavior

vscode-jest should update test result after error is fixed

### Actual Behavior

vscode-jest doesn't update test result when problem is fixed, it updates only on new errors.

### Other

I haven't worked on the project for a while, so I don't know when it stopped working, but it used to work fine before.

It looks like an ephemeral issue - after reproducing the steps for few dozen times it finally started working correctly. Still posting this issue in case anyone knows what is happening and why.

### PR DESCRIPTION
**Pull request by connectdotz**
part of the [v4](#576) initiative...

---

This PR introduces a new "interactive" mode that allows users to control when and what tests should be run, in addition to the current "watch" mode where jest/watcher will run the related tests whenever relevant files changes. This is a common pain point for projects with many tests and or a single source file change can trigger many test re-run, which can all cause high CPU, especially during the development phase. 

While working on this change, needed to refactor our process management model that we have wanted to do for a long time. Replaced the nested process model with a queue-based model, so no more keep-alive, callback, and assumption of the test sequence, etc. 

Another significant change is to split the `JestExt` into multiple files and move to a new directory. This helps better componentization and better test coverage. Hopefully will make future change easier as well.

The new process model added a few benefits that should make the extension smarter and liter not only for the interactive but the watch mode as well, such that we know what are test files without doing the full test run, and we can update snapshots without restarting the jest process. 

The last but not least, I think this PR is moving us closer to support the vscode's test API proposal,  which is an interactive test-item-driven provider framework. We still have some work to do and the API is not finalized yet, but it is clear that we do need to support an interactive mode anyway.

## Change Outline
- interactive mode
  - added a new setting `jest.autoRun` to let users control when the extension should run tests automatically. This supports: 
    1. the current "watcher" based test run. e.g.: `"jest.autoRun": {"watch": true, "onStartup": ["all-tests"]}`
    2. vscode file-change-event driven test run that can either run on only test file save or both test and source files. e.g.: `"jest.autoRun": {"watch": false, "onSave": "test-file"}` or `"jest.autoRun": {"watch": false, "onSave": "any-file", "onStartup": ["all-tests"]}` 
    3. the full manual mode where users trigger a test run via command or context-menu. e.g.: `{"jest.autoRun": "off"}`
  - user can trigger running the related tests for the current file from the editor context-menu:`Jest: Run Related Tests`. This menu will only appear in interactive mode and for both test and source files.
  - the coverage mode will obey the same autoRun scope.
  - The extension will invalidate the stale test result when the test file is changed but has not yet been run, or mark the project/workspace test stats are out-of-sync if the source/test file is changed but tests have not been run. 
  
- commands
  - added new commands for document-based jest commands, such as running a give test file, run all tests in the active workspace, toggle watch for the active workspace, etc.
  - remove the "Restart Runner" command. The "Start Runner" commands will restart if JestExt is already running.
  - adopt a more structured command id scheme to reflect the scope of the command.

- new context menu
  - added a context menu in the editor for running all tests in the file, which will only be available for non-watch interactive mode.

- new keyboard shortcut
  -  for the context menu 

- StatusBar
  - show workspace stats (success, fail, and unknown) for single and multi-root projects.
  - show autoRun mode
  - show workspace stats out of sync indicator when there are code changes that have not triggered related tests run.

- editor visual enhancement
  - remove debug code lens when the document is changed but tests have not been run to avoid interfering with the coding activity.
  - mark all tests unknown after the test file is saved but tests have not been run to accurately reflect the state of the tests.

- Snapshot enhancement
  - can update single file snapshot in the non-watch mode
  - updated snapshot will be immediately reflected without restarting the watch process.
  - fixed snapshot codeLens not showing (jest-community/jest-editor-support#63)
   
- refactor jest process management 
  -  removed any inter-process dependency, i.e. move from the nested process structure (all-tests followed by a watch run) to a queue-schedule model, which removed all the keepAlive, onExit callback, and other complicated state management logic.
  - removed any assumption about how the jest process should be started, it simply executes whatever is in the queue, which is scheduled by `JestExt` based on the `jest.autoRun`.
  - It supports both synchronous execution like before, i.e. watch won't start until all-test run completed; as well as concurrent runs so users can schedule updateSnapshot or other single file test run without blocked by a watch or other test runs. 
  - make JestProcess fully promise-based, the start and stop will return a promise that enables non-listener based interactions
  - support many more types of jest processes, such as getting all test files with "--listFiles", or running any test with `--updateSnapshot` flag. 

- refactor jest run handling / JestExt
  - split `JestExt.ts` into multiple files and move to a new folder. 
  - introduce a ProcessSession concept to explicitly bound test results/cache lifecycle.
  - take advantage of the new knowledge of the test files for stats tracking and reduce unnecessary test parsing. 

## Open Questions 
- debugCodeLens was using the file's basename for debugging, which could potentially cause multiple files being executed if they share the same file-name (under different directories). Don't see why we should not use the full file name, so I switched to the full fileName in the PR.  Please let me know if there is a legit reason otherwise.
- we used to re-register all the JestExt in `ExtensionManager` when config changed. This could lead to memory leaks as the previously registered JestExt were not being "deactivated". I did not see why they have to be registered again, so removed that logic altogether.  Please let me know if there is a legit reason otherwise.
- the new ProjectManager's non-blocking queue allows a max of 3 concurrent processes. Right now only the "onSave" autoRun uses this queue, so unless users do "save all", there should not be many concurrent use cases. I didn't make this configurable but if there is some performance issue, we could easily make it configurable.

## Testing
Added quite a bit of test along the way, which served as a live-testing for this PR. I did like the interactive mode with autoRun on test file save. Even though it starts a new process for every test file saves, at least on my mac, it still feels snappy compared to the watch mode. The big difference comes in when I change the source file like `JestExt.ts`, which means triggering 25 test files to be executed, not surprisingly a lot slower than just run `JestExt.test.ts` with interactive mode. But would be interested to see if how it performs in other OS/machine/projects...

Anyway, I am sure we need more testing and so probably earlier we can get people to start playing with it the better. I hope to cut a `jest-editor-support` release in the next couple of days so reviewers can run the actual extension during the review.

## ToDo
- [x] This PR does depend on `jest-editor-support` (jest-community/jest-editor-support#63), will update once it is released.
- [x] need to update README to document more detail of the new feature.

---
resolves #613 
resolves #668 
resolves #119
resolves #308
resolves #176 
resolves #408 
resolves #530 
resolves #36
resolves #574

### CODE REVIEW

**Comment by connectdotz** [comment]
I know this is a HUGE PR, sorry about that, trying to take advantage of the alpha status for this scale of changes. It is probably not feasible for one person to review everything, so I tagged a few reviewers. Please feel free to do a partial review for the part that you are most familiar with or interested in. 

In addition to code, I also made a pretty significant change to the README, would appreciate some editorial review and feedback.

@marcinczenko I think you are pretty familiar with the process-management, do you mind taking a look in that area? @stephtr would you mind looking at the JestExt and extension management? @orta and @seanpoulter feel free to jump in if you have some bandwidth...

File: README.md [hunk 1]
```diff
@@ -39,10 +52,6 @@ A comprehensive experience when using [Facebook's Jest](https://github.com/faceb
 
 <img src="https://github.com/jest-community/vscode-jest/raw/master/images/vscode-jest.gif" alt="Screenshot of the tool" width="100%">
 
-## Maintainers
-
-Orta Therox ([@orta](https://github.com/orta)), Vincent Voyer ([@vvo](https://github.com/vvo)) & ConnectDotz ([@connectdotz](https://github.com/connectdotz)).
```
Comment (orta) [inline README.md:44]: IMO, you should switch this to you 👍🏻

Comment (connectdotz) [reply README.md:44]: I just happened to have some freecycle recently so decided to take on these big-ticket items. Unfortunately, I didn't set a good example for the more "progressive" PR, but I think this kind of change is probably not a good fit for that. Nevertheless, I probably should have open a discussion thread before pulling the trigger... Please do feel free to voice your opinions, PR is not merged in, it's not too late. 

Ok, now to the actual point, I think we need more maintainers, not less, so we can have more people bouncing ideas and widening perspectives. How do you decide the maintainers? Do you have any suggestions on who else we can add? I know @stephtr is busy recently, but he has been very helpful in our recent PRs, do you think we could add him as the maintainer?

> IMO, you should switch this to you 👍🏻

You mentioned this couple of times, I might change my mind in the future, but right now I quite enjoy you being the owner so I have somebody to discuss and learn from 😄

**Comment by marcinczenko** [comment]
@connectdotz I will take a look. Most probably tomorrow.

**Comment by marcinczenko** [comment]
@connectdotz I could not find much time last week. Started today. I had to refresh myself a bit over the overall structure as I forgot most of the stuff ;). This is what I will be doing next: I am looking at processes-related stuff to see if I spot something that I feel needs some attention - not really deep - more in the spirit "can I follow it and does it make sense". I will also run it on some projects to see if I hit any troubles. I will try to reserve like an hour a day in the next week or so. I may also have some suggestions for the `README.md`.

In general the proposal looks good and it should improve the overall developer experience. In case I would be blocking anything please let me know.

**Comment by connectdotz** [comment]
@marcinczenko sounds good! Do what you can, any extra test and eyeballs is a plus, appreciate it! 🙏 

Aiming to start alpha testing next week if possible, let's see what we can get through before the end of the week...

**Comment by connectdotz** [comment]
@marcinczenko did you get a chance to test or review this PR further?

**Comment by connectdotz** [comment]
@orta I updated README to address your comment and tidy it up a bit. 

Looks like we are not going to get more testing or review by leaving the PR here... what do you think we should do? Should we merge it in and cut an alpha release so we can get a bigger community to try it?

**Comment by marcinczenko** [comment]
> @marcinczenko did you get a chance to test or review this PR further?

Yes, but it is going slow. I would explain myself better with a DM, but I do not know if my "whys" really matter. Shortly, I still need some time. Sorry about that... I like your comment about an alpha or a "canary" release so that we can win some time but already provide it for those who already want to give it a spin.

**Comment by orta** [comment]
The vscode marketplace doesn't really allow for separate versions of the same extension without having it as an entirely different app. You could se tup nightlies if you wanted, I did that for Svelte's vscode extension: https://github.com/sveltejs/language-tools/blob/master/.github/workflows/Deploy.yml

I'd double check that you're fine with it then ship it, and expect bug reports. It's OSS and in your spare time.

**Comment by connectdotz** [comment]
@orta, thanks for sharing the idea! 👍 

> I'd double check that you're fine with it then ship it, and expect bug reports. It's OSS and in your spare time.

True, I don't think it is ready for production yet, but keeping it here also does not get us closer to the finish line... Ideally, it would be better for people in this thread to test run it before open to the public. But it looks like everybody is super busy; therefore, I think a "private" alpha release might be an acceptable alternative to try moving forward. 
 
Let's do this: I will cut an alpha-5 release, including this PR and the wizard, but won't publish to the marketplace; announce it in the related issue threads to hopefully get more people testing it (it seems to work out ok with prior alpha releases).  

Meanwhile, @marcinczenko, thanks for trying to find time to review 👍 . I know how hard it is sometimes... Please continue and do not hesitate to ask any questions. After merging, this PR might be automatically closed, but you should still be able to review it. I will look out for your comments.

**Comment by marcinczenko** [comment]
@connectdotz Thanks for your understanding and patience. I will do my best.

**Pull request merged**